misk-aws/src/main/kotlin/misk/jobqueue/sqs/SqsJobConsumer.kt

```diff
@@ -45,91 +55,121 @@ internal class SqsJobConsumer @Inject internal constructor(
+      // Don't call handlers until all services are ready, otherwise handlers will crash because
+      // the services they might need (databases, etc.) won't be ready.
+      serviceManagerProvider.get().awaitHealthy()
+      receiver.runOnce()
     }
   }
 
   internal fun getReceiver(queueName: QueueName): QueueReceiver {
     return subscriptions[queueName]!!
   }
 
-  internal inner class QueueReceiver(
-    private val queueName: QueueName,
+  internal inner class QueueReceiver (
+    queueName: QueueName,
     private val handler: JobHandler
   ) {
     private val queue = queues[queueName]
 
     fun runOnce(): Status {
-      val (receiveDuration, messages) = queue.call { client ->
-        val receiveRequest = ReceiveMessageRequest()
-          .withAttributeNames("All")
-          .withMessageAttributeNames("All")
-          .withQueueUrl(queue.url)
-          .withMaxNumberOfMessages(config.message_batch_size)
-
-        timed { client.receiveMessage(receiveRequest).messages }
+      // Receive messages in parallel. Default to 1 if feature flag is not defined.
+      val numReceivers = max(featureFlags.getInt(RECEIVERS_PER_QUEUE, queue.name.value), 1)
+      val futures = (1..numReceivers)
+          .filter { num ->
+            val lease = leaseManager.requestLease("sqs-job-consumer-${queue.name.value}-$num")
+            config.clustered_consumers || lease.checkHeld()
+          }
+          .map {
+            CompletableFuture.supplyAsync(Supplier {
+              receive()
+            }, receivingThreads)
+          }
+
+      return CompletableFuture.allOf(*futures.toTypedArray()).thenApply {
+        futures.flatMap { it.join() }.fold(Status.OK) { acc, status ->
+          if (acc == Status.FAILED) {
+            acc
+          } else {
+            status
+          }
+        }
+      }.join()
+    }
+
+    private fun fetchMessages(): List<SqsJob> {
+      val messages = try {
+        metrics.sqsReceiveTime.timeCall(queue.queueName, queue.queueName) {
+          queue.call { client ->
+            val receiveRequest = ReceiveMessageRequest()
+                .withAttributeNames("All")
+                .withMessageAttributeNames("All")
+                .withQueueUrl(queue.url)
+                .withMaxNumberOfMessages(config.message_batch_size)
+
+            client.receiveMessage(receiveRequest).messages
+          }
+        }
+      } catch (e: ClientExecutionTimeoutException) {
+        log.info("timed out long polling for messages")
+        emptyList<Message>()
       }
 
-      val queueName = queue.name
+      return messages.map { SqsJob(queue.name, queues, metrics, it) }
+    }
 
-      metrics.sqsReceiveTime.record(receiveDuration.toMillis().toDouble(), queueName.value, queueName.value)
+    private fun receive(): List<Status> {
+      val messages = fetchMessages()
 
       if (messages.size == 0) {
-        return Status.NO_WORK
+        return listOf(Status.NO_WORK)
```

Comment:
I preserved the existing behavior here. However, I'm not sure if this is right. Do we want to back off if there are no messages? This could add some unexpected latency in the consumption of messages.

Reply:
yea you want to backoff, especially if the queue has little volume. ideally it would be configurable by queue